src/cpp/session/modules/rmarkdown/SessionRMarkdown.cpp

```diff
@@ -863,29 +870,42 @@ class RenderRmd : public async_r::AsyncRProcess
                           const std::string& output)
    {
       using namespace module_context;
-      if (type == module_context::kCompileOutputError &&  sourceNavigation_)
+      
+      if (type == kCompileOutputError && sourceNavigation_)
       {
-         // this is an error, parse it to see if it looks like a knitr error
-         const boost::regex knitrErr(
-                  "^Quitting from lines (\\d+)-(\\d+) \\(([^)]+)\\)(.*)");
-         boost::smatch matches;
-         if (regex_utils::match(output, matches, knitrErr))
+         if (renderErrorMarker_)
          {
-            // looks like a knitr error; compose a compile error object and
-            // emit it to the client when the render is complete
-            SourceMarker err(
-                     SourceMarker::Error,
-                     targetFile_.getParent().completePath(matches[3].str()),
-                     boost::lexical_cast<int>(matches[1].str()),
-                     1,
-                     core::html_utils::HTML(matches[4].str()),
-                     true);
-            knitrErrors_.push_back(err);
+            // if an error occurred during rendering, then any
+            // subsequent output should be gathered as part of
+            // that error
+            renderErrorMessage_ << output;
+         }
+         else
+         {
+            // check whether an error occurred while rendering the document and
+            // if knitr is about to exit. look for a specific quit marker and
+            // parse to to gather information for a source marker
+            const char* renderErrorPattern =
+                  "(?:.*?)Quitting from lines (\\d+)-(\\d+) \\(([^)]+)\\)(.*)";
+            
+            boost::regex reRenderError(renderErrorPattern);
+            boost::smatch matches;
+            if (regex_utils::match(output, matches, reRenderError))
+            {
+               // looks like a knitr error; compose a compile error object and
+               // emit it to the client when the render is complete
+               int line = boost::lexical_cast<int>(matches[1].str());
```

Comment:
lexical_cast can throw; I think we want safe_convert here (realize digits are guaranteed due to the regex, just hygiene)